File: SRC/statusbarhelper.cpp
#include "statusbarhelper.h"

#ifdef Q_OS_ANDROID
    #include <QJniObject>
    #include <QJniEnvironment>

StatusBarHelper::StatusBarHelper() {
}

float StatusBarHelper::getStatusBarHeight() {
    QJniObject activity = QJniObject::callStaticObjectMethod("org/qtproject/qt/android/QtNative",
                                                             "activity",
                                                             "()Landroid/app/Activity;");
    QJniObject statusBarHelper("com/raccoonline/vpnapp/StatusBarHelper",
                               "(Landroid/app/Activity;)V",
                               activity.object<jobject>());
    return statusBarHelper.callMethod<jfloat>("getStatusBarHeight");
}

bool StatusBarHelper::hasNotch() {
    QJniObject activity = QJniObject::callStaticObjectMethod("org/qtproject/qt/android/QtNative",
                                                             "activity",
                                                             "()Landroid/app/Activity;");
    QJniObject displayHelper("com/raccoonline/vpnapp/StatusBarHelper",
                             "(Landroid/app/Activity;)V",
                             activity.object<jobject>());
    return displayHelper.callMethod<jboolean>("hasNotch");
}

QRectF StatusBarHelper::getNotchSize() {
    QRectF     notchSize;
    QJniObject activity = QJniObject::callStaticObjectMethod("org/qtproject/qt/android/QtNative",
                                                             "activity",
                                                             "()Landroid/app/Activity;");
    QJniObject displayHelper("com/raccoonline/vpnapp/StatusBarHelper",
                             "(Landroid/app/Activity;)V",
                             activity.object<jobject>());
    QJniObject sizeArray = displayHelper.callObjectMethod("getNotchSize", "()[F");
    if (sizeArray.isValid()) {
        jfloatArray     array = sizeArray.object<jfloatArray>();
        QJniEnvironment env;
        jfloat         *sizes = env->GetFloatArrayElements(array, nullptr);
        notchSize             = QRectF(sizes[0], sizes[1], sizes[2], sizes[3]);
        env->ReleaseFloatArrayElements(array, sizes, 0);
    }
    return notchSize;
}

float StatusBarHelper::getNavigationBarHeight() {
    QJniObject activity = QJniObject::callStaticObjectMethod("org/qtproject/qt/android/QtNative",
                                                             "activity",
                                                             "()Landroid/app/Activity;");
    QJniObject statusBarHelper("com/raccoonline/vpnapp/StatusBarHelper",
                               "(Landroid/app/Activity;)V",
                               activity.object<jobject>());
    return statusBarHelper.callMethod<jfloat>("getNavigationBarHeight");
}

bool StatusBarHelper::hasNavigationBar() {
    QJniObject activity = QJniObject::callStaticObjectMethod("org/qtproject/qt/android/QtNative",
                                                             "activity",
                                                             "()Landroid/app/Activity;");
    QJniObject statusBarHelper("com/raccoonline/vpnapp/StatusBarHelper",
                               "(Landroid/app/Activity;)V",
                               activity.object<jobject>());
    return statusBarHelper.callMethod<jboolean>("hasNavigationBar");
}

float StatusBarHelper::getNavigationBarHeightForOrientation() {
    QJniObject activity = QJniObject::callStaticObjectMethod("org/qtproject/qt/android/QtNative",
                                                             "activity",
                                                             "()Landroid/app/Activity;");
    QJniObject statusBarHelper("com/raccoonline/vpnapp/StatusBarHelper",
                               "(Landroid/app/Activity;)V",
                               activity.object<jobject>());
    return statusBarHelper.callMethod<jfloat>("getNavigationBarHeightForOrientation");
}

bool StatusBarHelper::isGestureNavigationEnabled() {
    QJniObject activity = QJniObject::callStaticObjectMethod("org/qtproject/qt/android/QtNative",
                                                             "activity",
                                                             "()Landroid/app/Activity;");
    QJniObject statusBarHelper("com/raccoonline/vpnapp/StatusBarHelper",
                               "(Landroid/app/Activity;)V",
                               activity.object<jobject>());
    return statusBarHelper.callMethod<jboolean>("isGestureNavigationEnabled");
}

QList<float> StatusBarHelper::getSystemWindowInsets() {
    QList<float> insets;
    QJniObject   activity = QJniObject::callStaticObjectMethod("org/qtproject/qt/android/QtNative",
                                                             "activity",
                                                             "()Landroid/app/Activity;");
    QJniObject   statusBarHelper("com/raccoonline/vpnapp/StatusBarHelper",
                               "(Landroid/app/Activity;)V",
                               activity.object<jobject>());
    QJniObject   insetsArray = statusBarHelper.callObjectMethod("getSystemWindowInsets", "()[F");

    if (insetsArray.isValid()) {
        jfloatArray     array = insetsArray.object<jfloatArray>();
        QJniEnvironment env;
        jfloat         *values = env->GetFloatArrayElements(array, nullptr);

        // [left, top, right, bottom]
        for (int i = 0; i < 4; i++) {
            insets.append(values[i]);
        }

        env->ReleaseFloatArrayElements(array, values, 0);
    } else {
        // Fallback для старых версий
        insets.append(0.0f);                     // left
        insets.append(getStatusBarHeight());     // top
        insets.append(0.0f);                     // right
        insets.append(getNavigationBarHeight()); // bottom
    }

    return insets;
}

#else
StatusBarHelper::StatusBarHelper() {
}

float StatusBarHelper::getStatusBarHeight() {
    return 0.f;
}

bool StatusBarHelper::hasNotch() {
    return false;
}

QRectF StatusBarHelper::getNotchSize() {
    return QRectF();
}

float StatusBarHelper::getNavigationBarHeight() {
    return 0.f;
}

bool StatusBarHelper::hasNavigationBar() {
    return false;
}

float StatusBarHelper::getNavigationBarHeightForOrientation() {
    return 0.f;
}

bool StatusBarHelper::isGestureNavigationEnabled() {
    return false;
}

QList<float> StatusBarHelper::getSystemWindowInsets() {
    return QList<float> { 0.0f, 0.0f, 0.0f, 0.0f };
}
#endif

File: android/src/StatusBarHelper.java
package com.raccoonline.vpnapp;

import android.app.Activity;
import android.content.Context;
import android.view.WindowInsets;
import android.view.WindowManager;
import android.view.DisplayCutout;
import android.graphics.Rect;
import android.content.res.Configuration;

public class StatusBarHelper {

    private Activity activity;

    public StatusBarHelper(Activity activity) {
        this.activity = activity;
    }

    public float convertPxToDp(float px) {
        float density = activity.getResources().getDisplayMetrics().density;
        return px / density;
    }

    public float getStatusBarHeight() {
        int statusBarHeight = 0;

        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
            WindowInsets insets = activity.getWindow().getDecorView().getRootWindowInsets();

            if (insets != null) {
                statusBarHeight = insets.getStableInsetTop();
            }
        } else {
            // Используем устаревший метод для устройств до Android M
            int resourceId = activity.getResources().getIdentifier("status_bar_height", "dimen", "android");

            if (resourceId > 0) {
                statusBarHeight = activity.getResources().getDimensionPixelSize(resourceId);
            }
        }

        return convertPxToDp(statusBarHeight);
    }

    // Получить высоту навигационной панели
    public float getNavigationBarHeight() {
        int navigationBarHeight = 0;

        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
            WindowInsets insets = activity.getWindow().getDecorView().getRootWindowInsets();

            if (insets != null) {
                navigationBarHeight = insets.getStableInsetBottom();
            }
        } else {
            // Используем устаревший метод для устройств до Android M
            if (hasNavigationBar()) {
                int resourceId = activity.getResources().getIdentifier("navigation_bar_height", "dimen", "android");
                if (resourceId > 0) {
                    navigationBarHeight = activity.getResources().getDimensionPixelSize(resourceId);
                }
            }
        }

        return convertPxToDp(navigationBarHeight);
    }

    // Проверить, есть ли навигационная панель
    public boolean hasNavigationBar() {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
            WindowInsets insets = activity.getWindow().getDecorView().getRootWindowInsets();
            if (insets != null) {
                return insets.getStableInsetBottom() > 0;
            }
        }

        // Для старых версий Android используем проверку ресурсов
        int id = activity.getResources().getIdentifier("config_showNavigationBar", "bool", "android");
        if (id > 0) {
            return activity.getResources().getBoolean(id);
        }

        // Альтернативная проверка для некоторых устройств
        boolean hasMenuKey = android.view.ViewConfiguration.get(activity).hasPermanentMenuKey();
        boolean hasBackKey = android.view.KeyCharacterMap.deviceHasKey(android.view.KeyEvent.KEYCODE_BACK);
        
        return !hasMenuKey && !hasBackKey;
    }

    // Получить высоту навигационной панели в зависимости от ориентации
    public float getNavigationBarHeightForOrientation() {
        if (!hasNavigationBar()) {
            return 0;
        }

        String resourceName;
        int orientation = activity.getResources().getConfiguration().orientation;
        
        if (orientation == Configuration.ORIENTATION_LANDSCAPE) {
            resourceName = "navigation_bar_height_landscape";
        } else {
            resourceName = "navigation_bar_height";
        }

        int resourceId = activity.getResources().getIdentifier(resourceName, "dimen", "android");
        if (resourceId > 0) {
            return convertPxToDp(activity.getResources().getDimensionPixelSize(resourceId));
        }

        return 0;
    }

    // Проверить, используется ли жестовая навигация
    public boolean isGestureNavigationEnabled() {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.Q) {
            WindowInsets insets = activity.getWindow().getDecorView().getRootWindowInsets();
            if (insets != null) {
                // В жестовой навигации bottom inset обычно меньше обычной высоты навигационной панели
                int bottomInset = insets.getSystemWindowInsetBottom();
                int stableBottomInset = insets.getStableInsetBottom();
                
                // Если разница значительная, скорее всего используется жестовая навигация
                return Math.abs(bottomInset - stableBottomInset) > 20;
            }
        }
        return false;
    }

    public boolean hasNotch() {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.P) {
            WindowInsets windowInsets = activity.getWindow().getDecorView().getRootWindowInsets();

            if (windowInsets != null) {
                DisplayCutout displayCutout = windowInsets.getDisplayCutout();
                return displayCutout != null && !displayCutout.getBoundingRects().isEmpty();
            }
        }

        return false;
    }

    public float[] getNotchSize() {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.P) {
            WindowInsets windowInsets = activity.getWindow().getDecorView().getRootWindowInsets();

            if (windowInsets != null) {
                DisplayCutout displayCutout = windowInsets.getDisplayCutout();

                if (displayCutout != null && !displayCutout.getBoundingRects().isEmpty()) {
                    Rect rect = displayCutout.getBoundingRects().get(0);
                    float notchWidth = convertPxToDp(rect.right - rect.left); // Ширина выреза
                    float notchHeight = convertPxToDp(rect.bottom - rect.top); // Высота выреза
                    float x = convertPxToDp(rect.left); // X координата начала выреза
                    float y = convertPxToDp(rect.top); // Y координата начала выреза
                    return new float[] {
                        x, y, notchWidth, notchHeight
                    };
                }
            }
        }

        return new float[] {
            0, 0, 0, 0
        };
    }

    // Получить все отступы системы (статус бар, навигационная панель, боковые отступы)
    public float[] getSystemWindowInsets() {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
            WindowInsets insets = activity.getWindow().getDecorView().getRootWindowInsets();
            if (insets != null) {
                float left = convertPxToDp(insets.getSystemWindowInsetLeft());
                float top = convertPxToDp(insets.getSystemWindowInsetTop());
                float right = convertPxToDp(insets.getSystemWindowInsetRight());
                float bottom = convertPxToDp(insets.getSystemWindowInsetBottom());
                
                return new float[] { left, top, right, bottom };
            }
        }
        
        // Для старых версий возвращаем только статус бар и навигационную панель
        return new float[] { 
            0, 
            getStatusBarHeight(), 
            0, 
            getNavigationBarHeight() 
        };
    }
}
